### ISSUE TITLE
Translataion bug in tf.math.reduce_variance and tf.keras.backend.var

### ISSUE THREAD

**Issue by naolenikhil** [tensorflow/tensorflow#42694]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS High Sierra-10.13.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Not checked
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.8.5
- Bazel version (if compiling from source): 
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**Describe the current behavior**
Model built using tensorflow and later converter to tflite gives different results for same input. See code provided to reproduce the error. Issue [#42346](https://github.com/tensorflow/tensorflow/issues/42346), I created earlier, might have the same error.

**Command used to run the converter or code if you’re using the Python API**
Code to reproduce error with output
```
import tensorflow as tf
import tflite
import numpy as np
import imutils

tf.compat.v1.enable_eager_execution()

class MyFunc(tf.Module):
    @tf.function(input_signature=[tf.TensorSpec(shape=(None, None, 3), dtype=tf.float32, name='imdata')])
    def detect(self, imdata):
        # both tf.keras.backend.var and tf.math.reduce_variance operations create the same error scores
        #tensor_var = tf.keras.backend.var(imdata)
        tensor_var = tf.math.reduce_variance(imdata)
        return tensor_var
tf_func = MyFunc()

conc_func = tf_func.detect.get_concrete_function()
converter = tf.lite.TFLiteConverter.from_concrete_functions([conc_func])
fpath = 'model.tflite'
with tf.io.gfile.GFile(fpath, 'wb') as f:
    f.write(converter.convert())
loaded_model = tf.lite.Interpreter(fpath)

def tflite_inference(tmp_model, tmp_img_data):
    input_details = tmp_model.get_input_details()
    output_details = tmp_model.get_output_details()
    input_data = np.array(tmp_img_data).astype(input_details[0]['dtype'])
    w_img, h_img, ch_img = tmp_img_data.shape
    tmp_model.resize_tensor_input(input_details[0]["index"], [w_img, h_img, 3])
    tmp_model.allocate_tensors()
    tmp_model.set_tensor(input_details[0]['index'], tmp_img_data)
    tmp_model.invoke()
    result = tmp_model.get_tensor(output_details[0]['index'])
    return result


images = ['https://images.unsplash.com/photo-1470020337050-543c4e581988',
          'https://images.unsplash.com/photo-1516811108838-030371f93644',
          'https://zipbooks.com/wp-content/uploads/2017/05/royalty-free-images-free-of-charge.jpeg',
       ]

print("tensorflow vs tflite inference scores on images")
for img_url in images:
    tmp_imdata = imutils.url_to_image(img_url).astype(np.float32)
    print(f'\t tensorflow: {tf_func.detect(tmp_imdata).numpy():.4f}', end='\t')
    print(f'tflite: {tflite_inference(loaded_model, tmp_imdata):.4f}')

    
print("\n\n tensorflow vs tflite inference scores on random arrays")
shape = (2000, 2000, 3)
num_range = 255
arrays = [np.random.randint(num_range, size=shape).astype(np.float32), 
          np.random.randint(num_range, size=shape).astype(np.float32),
          np.random.randint(num_range, size=shape).astype(np.float32)
         ]
print('\t tensorflow \t tflite \t   Equal?')
for arr in arrays:
    tf_score = tf_func.detect(arr).numpy()
    tflite_score = tflite_inference(loaded_model, arr)
    eq = tf_score == tflite_score
    print(f'\t {tf_score:.4f} \t {tflite_score:.4f} \t   {eq}')
```

**The output from the converter invocation** : 
```
tensorflow vs tflite inference scores on images
	 tensorflow: 4722.1460	tflite: 4706.3296
	 tensorflow: 5759.6831	tflite: 4486.9917
	 tensorflow: 1643.2089	tflite: 1642.7393


 tensorflow vs tflite inference scores on random arrays
	 tensorflow 	 tflite 	   Equal?
	 5417.6357 	 5351.8447 	   False
	 5418.3472 	 5352.1660 	   False
	 5419.2515 	 5353.3892 	   False
```

**Also, please include a link to the saved model or GraphDef** : Unzip file [model.tflite.zip](https://github.com/tensorflow/tensorflow/files/5133550/model.tflite.zip) .

**Failure details**
If the conversion is successful, but the generated model is wrong,
state what is wrong:
- Producing wrong results and/or decrease in accuracy: Wrong results
- Producing correct results, but the model is slower than expected (model generated from old converter): No

**Any other info / logs**
See above for source code and logs.

**Comment by ravikyram** [comment]
I have tried in colab with TF versions 2.3, nightly versions and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/5b490846517f4d553c747ddc00cd83d7/untitled272.ipynb).Thanks!

**Comment by srjoglekar246** [comment]
There are three ops in the execution plan for the (tflite) model attached: MEAN, SQUARED_DIFFERENCE, MEAN. These come from the TF implementation of [reduce_variance](https://www.tensorflow.org/api_docs/python/tf/math/reduce_variance).

I tried playing around with @ravikyram 's gist, and looks like the accuracy issue might stem from TFLite's reduction ops. I modified the script to use reduce_sum instead of reduce_variance:

```
class MyFunc(tf.Module):
    @tf.function(input_signature=[tf.TensorSpec(shape=(None, None, 3), dtype=tf.float32, name='imdata')])
    def detect(self, imdata):
        tensor_var = tf.math.reduce_sum(imdata)
        return tensor_var
tf_func = MyFunc()
```

In this case, for the random inputs generated at the end of the script, TF & TFLite outputs match upto a certain dim size (200, 200, 3):

```
print("\n\n tensorflow vs tflite inference scores on random arrays")
shape = (200, 200, 3)
num_range = 255
arrays = [np.random.randint(num_range, size=shape).astype(np.float32), 
          np.random.randint(num_range, size=shape).astype(np.float32),
          np.random.randint(num_range, size=shape).astype(np.float32)
         ]
print('\t tensorflow \t tflite \t   Equal?')
for arr in arrays:
    tf_score = tf_func.detect(arr).numpy()
    tflite_score = tflite_inference(loaded_model, arr)
    eq = tf_score == tflite_score
    print(f'\t {tf_score:.4f} \t {tflite_score:.4f} \t   {eq}')
``` 

However, for larger float32 inputs, there is a difference between what TF & TFLite provide.
The dip in accuracy gradually grows from (say) 230x230x3 to 500x500x3, and becomes much larger for the dimensions tried by @naolenikhil (2000x2000x3).

reduce_min & reduce_max don't have this issue, so there might be some problem with the accumulators used by TFLite kernels?

**Comment by srjoglekar246** [comment]
@talumbau Would you have an idea of what might be the issue here? (Refer comment above)

**Comment by grantjensen** [comment]
This is due to the imprecision of floats at high values. In tflite, we perform reduce_sum on a float32 input by summing the inputs with float32 operations. However, TensorFlow will do the math as float64 and then cast back. 
One example you can try yourself is to perform 33,554,440 + 10 with float32. Clearly the answer should be 33,554,450. However, if we look at some IEEE 754 conversions (https://www.h-schmidt.net/FloatConverter/IEEE754.html): 

- 33,554,440 := 0x4c000002; 
- 33,554,444 := 0x4c000003; 
- 33,554,448 := 0x4c000004; 
- 33,554,452 := 0x4c000005;

This is the finest precision that floats can represent at such large values. This is intendend! In general, floats have great precision at low values, and bad precision at high values. Ie, 33,554,450 cannot be represented exactly in IEEE 754, so instead we get the value 33,554,448. These small rounding errors add up over time, resulting in the mismatches shown in your bug.

In short, this is expected behavior from summing float32 values; try your best to form your model to avoid summing such large numbers.
